docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,223 @@
+		// wait for dependency to complete build
+		select {
+		case <-ctx.Done():
+		case <-dep:
+		}
+	}
+}
+```
+
+Each artifact has a channel that it closes once it completes building (either success or failure) by calling `markComplete`. This notifies *all* listeners waiting for this artifact.
+
+Additionally it has a reference to the channels for each of its dependencies.
+Calling `waitForDependencies` ensures that all required artifacts' channels have already been closed and as such have finished building before the current artifact build starts.
+
+This class also provides an implementation of the interface `buildStatusRecorder` that should be safe for concurrent access.
+
+```go
+type buildStatusRecorder interface {
+  Record(imageName string, imageTag string, err error)
+  GetImageTag(imageName string) string
+}
+```
+
+Finally we have the only exported function in `scheduler.go` that orchestrates all the builds:
+
+```go
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, existing []Artifact,  buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
+```
+
+This function maintains an instance of `buildStatusRecorder` implementation and can pass it as an `ArtifactResolver` to the various `ArtifactBuilder`s while recording the status after each build completion.
+
+## Build logs reporting
+
+There are two quirks in the way the current `InParallel` build controller reports the build logs.
+
+```go
+func collectResults(out io.Writer, artifacts []*latest.Artifact, results *sync.Map, outputs []chan string) ([]Artifact, error) {
+	var built []Artifact
+	for i, artifact := range artifacts {
+		// Wait for build to complete.
+		printResult(out, outputs[i])
+		v, ok := results.Load(artifact.ImageName)
+		if !ok {
+			return nil, fmt.Errorf("could not find build result for image %s", artifact.ImageName)
+		}
+		switch t := v.(type) {
+		case error:
+			return nil, fmt.Errorf("couldn't build %q: %w", artifact.ImageName, t)
+		case Artifact:
+			built = append(built, t)
+		default:
+			return nil, fmt.Errorf("unknown type %T for %s", t, artifact.ImageName)
+		}
+	}
+	return built, nil
+}
+```
+
+- It reports in the order of artifacts in the `Artifact` slice instead of the actual order in which they get built. 
+- It only reports a single artifact build failure even though there could have been multiple failures.
+
+This will prove misleading after the introduction of artifact dependencies since we can have out of order artifact definitions in the skaffold config which with the current reporting strategy would appear to be building in the wrong order, and also build failures due to failed required artifact builds won't be immediately apparant.
+
+So we introduce a new `BuildLogger` that is both order aware while at the same time does not interleave log messages due to some builds running concurrently. Additionally it concats all build failure messeges and reports all of them.
```

Comment:
Do you have the details of this scoped somewhere?
I would suggest, we first implement
1. Artifact dependency in sequential order. Meaning a single artifact will be built at a time. Keep in mind, we design for both running in sequence an in parallel. However, in-sequence builds will be faster to land.
2. Design UX for running builds in parallel
3. Switch to running builds concurrently.

Reply:
It makes sense to design the UI and then add functionality execute builds concurrently to make sure we add features with good UX.

WDYT?

Reply:
Add a section of "Future work" and add "Re-desing UX for concurrent builds"

Reply:
Added your suggestions